### PR TITLE
IEN refactor Applicant, Status and audit details

### PR DESCRIPTION
**Pull request by bhumin-fw**

### CODE REVIEW

File: apps/api/src/applicant/dto/ienapplicant-create.dto.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  @ApiProperty({ description: 'Applicant Name', default: 'Mark Bowlill' })
+  @IsString()
+  @Length(1, 256, { message: 'Please provide applicant name' })
+  name!: string;
+
+  @ApiPropertyOptional({
+    description: 'Applicant unique ID',
+    default: 'a989765-c8789299-b9696669-dfcba98',
+  })
+  @Length(1, 256, { message: 'Please provide applicant name' })
```
Comment (coryatwater-FW) [inline apps/api/src/applicant/dto/ienapplicant-create.dto.ts:23]: I think a typo here for the message?

File: apps/api/src/applicant/dto/ienapplicant-update.dto.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  @Length(1, 256, { message: 'Please provide applicant name' })
+  name?: string;
+
+  @ApiPropertyOptional({
+    description: 'Applicant unique ID',
+    default: 'a989765-c8789299-b9696669-dfcba98',
+  })
+  @Length(1, 256, { message: 'Please provide applicant name' })
+  @IsOptional()
+  applicant_id?: string;
```
Comment (coryatwater-FW) [inline apps/api/src/applicant/dto/ienapplicant-update.dto.ts:25]: same typo here

File: apps/api/src/applicant/dto/ienapplicant-update.dto.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  IsOptional,
+  IsString,
+  Length,
+  IsObject,
+  IsBoolean,
+  IsArray,
+} from 'class-validator';
+import { IENApplicantUpdateDTO } from '@ien/common';
+
+export class IENApplicantUpdateAPIDTO extends IENApplicantUpdateDTO {
```
Comment (dbayly-freshworks) [inline apps/api/src/applicant/dto/ienapplicant-update.dto.ts:13]: Duplicating all these DTO's in order to add documentation feels like an anti pattern, we should create a task to see if we can fix this. There should be a better way to do this.

File: apps/api/src/applicant/dto/ienapplicant-create.dto.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  name!: string;
+
+  @ApiPropertyOptional({
+    description: 'Applicant unique ID',
+    default: 'a989765-c8789299-b9696669-dfcba98',
+  })
+  @Length(1, 256, { message: 'Please provide applicant unique ID' })
+  @IsOptional()
+  applicant_id?: string;
```
[the remark this answers is not in the document]
Comment (bhumin-fw) [reply apps/api/src/applicant/dto/ienapplicant-create.dto.ts:25]: @dbayly-freshworks 
We have to keep it(applicant_id that already generated in HMBC ATS system) open, based on this ID and unique key on it, we will be able to upsert the record in a bulk manner.

All open fields other than "education" are somewhat used in the Where clause to filter or join another table.

File: apps/api/src/applicant/dto/ienapplicant-update.dto.ts (hunk 3 above)
Comment (bhumin-fw) [reply apps/api/src/applicant/dto/ienapplicant-update.dto.ts:13]: I'd share one link in our chat about this issue discuss on git, but was not able to resolve it in nextjs FE, It works fine with the decorator in nestjs. So I had put it as it is, If you get some time could you please take a look at it this or maybe next week.

File: apps/api/src/applicant/ienapplicant.controller.ts [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+  //   @Body(new ParseArrayPipe({ items: ApplicantCreateAPIDTO }))
+  //   applicants: ApplicantCreateAPIDTO[],
+  // ) {
+  //   try {
+  //     return this.applicantService.createBulkApplicants(applicants);
+  //   } catch (e) {
+  //     this.logger.error(e);
+  //     throw new InternalServerErrorException('An unknown error occured while update applicant');
+  //   }
+  // }
```
[the remark this answers is not in the document]
Comment (bhumin-fw) [reply apps/api/src/applicant/ienapplicant.controller.ts:166]: I have keep it for the bulk upload function, Let me remove it for now.

File: apps/api/src/applicant/dto/ienapplicant-by-id.dto.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+export class IENApplicantFilterByIdAPIDTO extends IENApplicantFilterByIdDTO {
+  @ApiPropertyOptional({
+    description:
+      'Get additional data, available values are [audit, applicantaudit]. Pass it as comma separated like audit or audit,applicantaudit or applicantaudit',
+  })
+  @IsNotEmpty()
+  @Length(1, 256)
+  @IsOptional()
+  @IsString()
```
[the remark this answers is not in the document]
Comment (bhumin-fw) [reply apps/api/src/applicant/dto/ienapplicant-by-id.dto.ts:13]: Accepting a string with comma-separated values. It is a GET API, and the param will pass on Query string only. So do not want to add any array here. For now, we have only 2 options for extra details in the future maybe we extend it.
audit and applicantaudit. you can pass a single value or a comma-separated one.

**Review by dbayly-freshworks — APPROVED** [review]

**Pull request merged**